File: client/src/types/todo.ts
// Todo types
export interface Todo {
  id: string;
  title: string;
  description?: string;
  due_time?: string;
  is_completed: boolean;
  is_important: boolean;
  user_id?: string;
  group_id?: string;
  created_at: string;
  updated_at: string;
  tags?: Tag[];
}

export interface TodoCreateRequest {
  title: string;
  description?: string;
  due_time?: string;
  group_id?: string;
  tag_ids?: string[];
  is_important?: boolean;
}

export interface TodoUpdateRequest {
  title?: string;
  description?: string;
  due_time?: string;
  group_id?: string;
  is_important?: boolean;
  is_completed?: boolean;
}

// Group types
export interface Group {
  id: string;
  name?: string;
  created_at: string;
  updated_at: string;
}

export interface GroupCreateRequest {
  name: string;
}

export interface GroupUpdateRequest {
  name?: string;
}

// Tag types
export interface Tag {
  id: string;
  name?: string;
  created_at: string;
  updated_at: string;
}

export interface TagCreateRequest {
  name: string;
}

export interface TagUpdateRequest {
  name?: string;
}

// Filter types
export interface TodoFilters {
  is_completed?: boolean;
  is_important?: boolean;
  q?: string;
  group_id?: string;
  tag_id?: string;
  limit?: number;
  offset?: number;
}

File: server/app/models/todo.py
from datetime import datetime
from typing import List, Optional

import sqlalchemy as sa
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy import ForeignKey
from sqlalchemy.orm import Mapped, mapped_column, relationship

from . import Base


class Todo(Base):
    __tablename__ = "todos"

    id: Mapped[str] = mapped_column(
        UUID(as_uuid=False),
        primary_key=True,
        server_default=sa.text("uuid_generate_v4()"),
    )
    user_id: Mapped[Optional[str]] = mapped_column(
        UUID(as_uuid=False), ForeignKey("users.id"), nullable=True
    )
    group_id: Mapped[Optional[str]] = mapped_column(
        UUID(as_uuid=False), ForeignKey("groups.id"), nullable=True
    )
    title: Mapped[str] = mapped_column(sa.Text, nullable=False)
    description: Mapped[Optional[str]] = mapped_column(sa.Text, nullable=True)
    due_time: Mapped[Optional[datetime]] = mapped_column(sa.DateTime(timezone=True), nullable=True)
    is_completed: Mapped[bool] = mapped_column(sa.Boolean, nullable=False, server_default=sa.text("false"))
    is_important: Mapped[bool] = mapped_column(sa.Boolean, nullable=False, server_default=sa.text("false"))
    created_at: Mapped[datetime | None] = mapped_column(
        sa.DateTime(timezone=True), server_default=sa.text("now()"), nullable=True
    )
    updated_at: Mapped[datetime | None] = mapped_column(
        sa.DateTime(timezone=True), server_default=sa.text("now()"), nullable=True
    )

    user: Mapped[Optional["User"]] = relationship("User", foreign_keys=[user_id], back_populates="todos")
    # relationships
    group: Mapped[Optional["Group"]] = relationship("Group")
    notifications = relationship("Notification", back_populates="todo")
    
    # Many-to-many relationship with tags through TodoTag
    tags = relationship(
        "Tag",
        secondary="todo_tag",
        primaryjoin="Todo.id == TodoTag.todo_id",
        secondaryjoin="Tag.id == TodoTag.tag_id",
        back_populates="todos"
    )


